test_conformance/basic/test_vector_swizzle.cpp

```diff
@@ -0,0 +1,681 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <algorithm>
+#include <numeric>
+#include <string>
+#include <vector>
+
+#include "procs.h"
+#include "harness/testHarness.h"
+
+template <int N> struct TestInfo
+{
+};
+
+template <> struct TestInfo<2>
+{
+    static const size_t vector_size = 2;
+
+    static constexpr const char* kernel_source_xyzw = R"CLC(
+__kernel void test_vector_swizzle_xyzw(TYPE value, __global TYPE* dst) {
+    int index = 0;
+
+    // lvalue swizzles
+    dst[index++].x = value.x;
+    dst[index++].y = value.x;
+    dst[index++].xy = value;
+    dst[index++].yx = value;
+
+    // rvalue swizzles
+    dst[index++] = value.x;
+    dst[index++] = value.y;
+    dst[index++] = value.xy;
+    dst[index++] = value.yx;
+}
+)CLC";
+
+    static constexpr const char* kernel_source_rgba = R"CLC(
+__kernel void test_vector_swizzle_rgba(TYPE value, __global TYPE* dst) {
+    int index = 0;
+
+    // lvalue swizzles
+    dst[index++].r = value.r;
+    dst[index++].g = value.r;
+    dst[index++].rg = value;
+    dst[index++].gr = value;
+
+    // rvalue swizzles
+    dst[index++] = value.r;
+    dst[index++] = value.g;
+    dst[index++] = value.rg;
+    dst[index++] = value.gr;
+}
+)CLC";
+
+    static constexpr const char* kernel_source_sN = R"CLC(
+__kernel void test_vector_swizzle_sN(TYPE value, __global TYPE* dst) {
+    int index = 0;
+
+    // lvalue swizzles
+    dst[index++].s0 = value.s0;
+    dst[index++].s1 = value.s0;
+    dst[index++].s01 = value;
+    dst[index++].s10 = value;
+
+    // rvalue swizzles
+    dst[index++] = value.s0;
+    dst[index++] = value.s1;
+    dst[index++] = value.s01;
+    dst[index++] = value.s10;
+}
+)CLC";
+};
+
+template <> struct TestInfo<3>
+{
+    static const size_t vector_size = 4; // sizeof(vec3) is four elements
+
+    static constexpr const char* kernel_source_xyzw = R"CLC(
+__kernel void test_vector_swizzle_xyzw(TYPE value, __global TYPE* dst) {
+    int index = 0;
+
+    // lvalue swizzles
+    dst[index++].x = value.x;
+    dst[index++].y = value.x;
+    dst[index++].z = value.x;
+    dst[index++].xyz = value;
+    dst[index++].zyx = value;
+
+    // rvalue swizzles
+    vstore3(value.x, 0, (BASETYPE*)(dst + index++));
```

Comment:
I think for devices that do not support generic adress space all these `(BASETYPE*)` casts should have a `__global` because `dst` is `__global`?

```suggestion
    vstore3(value.x, 0, (__global BASETYPE*)(dst + index++));
```
I'm seeing errors such as

```
kernel.opencl:13:25: error: casting '__global char3 *' to type '__private char *' changes address space of pointer
    vstore3(value.x, 0, (BASETYPE*)(dst + index++));
                        ^          ~~~~~~~~~~~~~~~
```

Reply:
Good catch - I've fixed all of these.

Reply:
Thanks!